src/Octokit.Webhooks/WebhookUtilities.cs

```diff
@@ -0,0 +1,113 @@
+                this.StatusBody = statusBody;
+            }
+        }
+
+        private static VerifyResult<bool> VerifyContentType(HttpRequest req, string expected)
+        {
+            if (expected == req.ContentType)
+            {
+                return new VerifyResult<bool>(true);
+            }
+            return new VerifyResult<bool>(HttpStatusCode.BadRequest, "Unexpected content type");
+        }
+
+        private static async Task<string> GetBodyAsync(HttpRequest request)
+        {
+            using var reader = new StreamReader(request.Body);
+            return await reader.ReadToEndAsync().ConfigureAwait(false);
+        }
+
+        private static async Task<VerifyResult<string>> VerifySignature(HttpRequest request, string secret)
+        {
+            var body = await GetBodyAsync(request).ConfigureAwait(false);
+            request.Headers.TryGetValue("X-Hub-Signature-256", out var signatureSha256);
+
+            var isSigned = signatureSha256.Count > 0;
+            var isSignatureExpected = !string.IsNullOrEmpty(secret);
+
+            if (!isSigned && !isSignatureExpected)
+            {
+                return new VerifyResult<string>(body);
+            }
+
+            if (!isSigned && isSignatureExpected)
+            {
+                return new VerifyResult<string>(HttpStatusCode.BadRequest, "Missing signature header");
+            }
+
+            if (isSigned && !isSignatureExpected)
+            {
+                return new VerifyResult<string>(HttpStatusCode.InternalServerError,
+                    "Payload includes a secret, so the webhook receiver must configure a secret.");
+            }
+
+            var keyBytes = Encoding.UTF8.GetBytes(secret);
+            var bodyBytes = Encoding.UTF8.GetBytes(body);
+            var hmac = new HMACSHA256(keyBytes);
+            var hash = hmac.ComputeHash(bodyBytes);
+            var hashHex = Convert.ToHexString(hash);
+            var expectedHeader = $"sha256={hashHex.ToLower(CultureInfo.InvariantCulture)}";
+            if (signatureSha256.ToString() != expectedHeader)
+            {
+                return new VerifyResult<string>(HttpStatusCode.BadRequest, "Signature mismatch");
+            }
+
+            return new VerifyResult<string>(body);
+        }
+
+        public static async Task<VerifyResult<string>> GetBodyChecked(HttpRequest request, string secret)
+        {
+            // Verify content type
+            var vct = VerifyContentType(request, MediaTypeNames.Application.Json);
+            if (vct.StatusCode != HttpStatusCode.OK)
+            {
+                return new VerifyResult<string>(vct.StatusCode, vct.StatusBody);
+            }
+
+            // Require a length up front and that it not be too long
+            if (request.ContentLength == 0)
+            {
+                return new VerifyResult<string>(HttpStatusCode.LengthRequired, String.Empty);
+            }
+
+            if (request.ContentLength > 1024 * 1024)
```

Comment:
https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#webhook-payload-object-common-properties

> Payloads are capped at 25 MB.